### PR TITLE
added minimal c# wrapper dll and c# example project

### PR DESCRIPTION
**Pull request by digital-synapse**
not sure if you were looking for additional language bindings but here is the start to the C api wrapper with enough support to run the example project in c# on mono

### CODE REVIEW

**Comment by hzeller** [comment]
Cool thanks!

Before we can do the merging, can you try to minimize that ?
  * the gitignore blew up significantly in size, but I suspect there are only really a handful new file types if at all that need to be added.
  * Please don't use visual studio and its project file (it is not open source, so it would force people to use a proprietary tool which they might not have access to), but instead write a Makefile that does the same thing. It should entirely be possible to built directly on the Pi with a Makefile. Also, you'll see that simplifies everything significantly.

**Comment by digital-synapse** [comment]
sure i can add a makefile and reduce the gitignore. not a problem, but are you sure you want to exclude the project file? VS is by far the most common IDE for c#

**Comment by hzeller** [comment]
Yes, Makefile is enough. You reduce a complexity to a few lines that is easy to understand and read.

If someone then wants to use the resulting library in Visual Studio they are still free to integrate it in their build system.

**Comment by hzeller** [comment]
Also, the assembly info files don't really look like they are necessary ?

File: c#/RGBLedMatrix.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace rpi_rgb_led_matrix_sharp
+{
+    public class RGBLedMatrixSharp
+    {
+        [DllImport("librgbmatrix.so")]
+        public static extern IntPtr led_matrix_create(int rows, int chained, int parallel);
```
Comment (hzeller) [inline c#/RGBLedMatrix.cs]: This IntPtr looks like it is not typed, so it is easy to accidentally use wrong.
Is there a way that you can create dummy types so that the pointers that represent a Matrix and a Canvas are different ? I don't know who that is done is CS, so I let it up to you to choose the best way to do that.

File: c#/RGBLedMatrix.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        public static extern void led_canvas_set_pixel(IntPtr canvas, int x, int y, byte r, byte g, byte b);
+
+        [DllImport("librgbmatrix.so")]
+        public static extern void led_canvas_clear(IntPtr canvas);
+
+        [DllImport("librgbmatrix.so")]
+        public static extern void led_canvas_fill(IntPtr canvas, byte r, byte g, byte b);
+
+        [DllImport("librgbmatrix.so")]
+        public static extern IntPtr led_matrix_create_offscreen_canvas(IntPtr matrix);
```
Comment (hzeller) [inline c#/RGBLedMatrix.cs]: Aren't we missing led_matrix_get_canvas() ?

**Comment by digital-synapse** [comment]
good to go.
added makefile, removed all the visual studio files and cleaned up the gitignore

**Comment by digital-synapse** [comment]
sorry for all the commits - was trying to clear out the few remaining unneeded files.

**Comment by hzeller** [comment]
Cool, this looks much better. I'll have a closer look tomorrow, its getting night here.

Is there a way to use something like `LD_LIBRARY_PATH` or the equivalent of `-rpath` with csharp programs to help it find the dll ? Right now it looks like everything has to be copied together in a directory for it to work.

Don't worry about all the commits, I'll be squashing them nicely together.

**Comment by digital-synapse** [comment]
As far as i know mono and windows applications always just have their dlls
in the same folder as the executable. However mono does have a concept of a
GAC which may offer somthing like that but i have not used it myself
http://www.mono-project.com/docs/advanced/assemblies-and-the-gac/

**Comment by digital-synapse** [comment]
I looked at adding support for the DrawFont methods but they look quite a bit more involved.

**Comment by digital-synapse** [comment]
Pull request ready for review :)

File: c#/testapp/App.config [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
```
Comment (hzeller) [inline c#/testapp/App.config]: Is this file needed ?

File: c#/RGBLedMatrix.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        #region IDisposable Support
+        private bool disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
```
Comment (hzeller) [inline c#/RGBLedMatrix.cs]: So is this something needed and you intend to implement something here, or can this whole if() go ?

File: c#/Makefile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+all:
```
Comment (hzeller) [inline c#/Makefile]: This Makefile rule does not make sure that
  (a) the librgbmatrix.so is build. You need to depend on that
  (b) does not do anything depending on the change of the cs files. So you want to depend on each.
  (c) the test program is part of building the library, which is probably what you want to keep independent.

So you probably want to have something like this
```
CSHARP_LIB=RGBLedMatrix.dll
SOURCES=RGBLedCanvas.cs RGBLedMatrix.cs
CSHARP_COMPILER=mcs

RGB_LIBDIR=../lib
RGB_LIBRARY_NAME=rgbmatrix
RGB_LIBRARY=$(RGB_LIBDIR)/lib$(RGB_LIBRARY_NAME).so.1

$(CSHARP_LIB) : $(SOURCES) $(RGB_LIBRARY)
        $(CSHARP_COMPILER) -target:library -out:$@ $(SOURCES)

$(RGB_LIBRARY):
        $(MAKE) -C $(RGB_LIBDIR)
```

for the library, and then an independent target for the test program.

File: .gitignore [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
+*.csproj
+[Dd]ebug/
+[Rr]elease/
+x64/
+x86/
+[Bb]in/
+[Oo]bj/
+[Ll]og/
+.vs/
+AssemblyInfo.cs
```
Comment (hzeller) [inline .gitignore]: Is this a generated file or why does it show up here ?

File: .gitignore [hunk 7]
```diff
@@ -1,7 +1,22 @@
 *.o
 *.a
 *~
+*.so
 led-image-viewer
 demo
 minimal-example
 text-example
+
+# visual studio
```
Comment (hzeller) [inline .gitignore]: All these files don't look like they would be generated if the Makefile in this project is being used, so we don't need to mention them here. Probably the only file we need to mention is *.dll

File: Makefile [hunk 8]
```diff
@@ -29,6 +30,9 @@ clean:
 	$(MAKE) -C utils clean
 	$(MAKE) -C $(PYTHON_LIB_DIR) clean
 
+build-csharp:
+	$(MAKE) -C $(CSHARP_LIB_DIR) build
```
Comment (hzeller) [inline Makefile]: You don't have a `build` target in your Makefile. So you might either add it there or mention the target `all` that is in the Makefile.

**Comment by hzeller** [comment]
Cool, will have a look later tonight, here is a first round of smallish comments.

Can you add a README.md similar to the one in python/ that describes what needs to be installed (I suspect `mono-mcs`), point out how to run the test program and generalize from there how someone would get started doing their on program (e.g. mention that it is necessary to copy the libraries into the directory of the binary).

To provide more flexibility, it would actually be cool if people could pass something comparable to`struct RGBLedMatrixOptions` as it allows more flexibility than just rows/parallel/chain. But that can as well be a separate pull request.
You then can just `led_matrix_create_from_options()` with an options struct and leave *argc and **argv to NULL.

File: c#/testapp/App.config (hunk 3 above)
Comment (digital-synapse) [inline c#/testapp/App.config]: no it can be removed

File: c#/RGBLedMatrix.cs (hunk 4 above)
Comment (digital-synapse) [inline c#/RGBLedMatrix.cs]: this was generated the empty if () can be removed

File: c#/Makefile (hunk 5 above)
Comment (digital-synapse) [inline c#/Makefile]: looks good to me. question. why is the library *.so.1 instead of just *.so? I was confused about this

File: .gitignore (hunk 6 above)
Comment (digital-synapse) [inline .gitignore]: yes, these folders and files are generated by visual studio. I added them to the gitignore to make it easier for people using VS to contribute. They can be removed from the gitignore if desired.

File: Makefile (hunk 8 above)
Comment (digital-synapse) [inline Makefile]: thanks

**Comment by digital-synapse** [comment]
sure. yeah i think mono-mcs would probably do it for the demo. I'll add a readme with some info.

**Comment by digital-synapse** [comment]
will add the RGBLedMatrixOptions in a seperate pull request.

**Comment by hzeller** [comment]
Thanks!
